modules/fedimint-wallet-client/src/backup/recovery_history_tracker.rs

```diff
@@ -0,0 +1,128 @@
+use std::collections::{BTreeMap, VecDeque};
+
+use fedimint_core::encoding::{Decodable, Encodable};
+use fedimint_logging::LOG_CLIENT_MODULE_WALLET;
+use tracing::debug;
+
+use super::RECOVER_MAX_GAP;
+use crate::client_db::TweakIdx;
+use crate::WalletClientModuleData;
+
+/// Tracks addresses `TweakIdx`s/addresses that are expected to have been used
+/// against the stream of addresses that were actually used for peg-ins in the
+/// Federation.
+///
+/// Since replaying Federation history is entirely private, the goal here
+/// is to find the last peg-in address already used without compromising
+/// privacy like when querying Bitcoin node.
+///
+/// While at it, collect some addresses that were actually used for peg-ins by
+/// other clients, just to query for them as decoys and thus hopefully make the
+/// malicious Bitcoin node operator have less confidence about which addresses
+/// are actually linked with each other.
+#[derive(Clone, Debug, Encodable, Decodable)]
+pub struct ConsensusPegInTweakIdxesUsedTracker {
+    pending_pubkey_scripts: BTreeMap<bitcoin::ScriptBuf, TweakIdx>,
+    next_pending_tweak_idx: TweakIdx,
+    next_unused_tweak_idx: TweakIdx,
+    decoys: VecDeque<bitcoin::ScriptBuf>,
+    // To avoid updating `decoys` for the whole recovery, which might be a lot of extra updates
+    // most of which will be thrown away, ignore script pubkeys from before this `session_idx`
+    decoy_session_threshold: u64,
```

Comment:
The cost of a cloning is more of a problem, potentially. Also, we really only want some recently used addresses as a decoy, as we are going to check only few (potentially) last used addresses. An attacker might see through things if we query some much older addresses.

Reply:
Since I might need to query old addresses anyway, maybe I'll just replace it with some probabilistic thing to potentially keep some older addresses around...